### ISSUE TITLE
Closing GUI After Wallet Open for a While Results in Infinite Loading Animation 

### ISSUE THREAD

**Issue by elibroftw** [monero-project/monero-gui#3840]
OS: Windows

If I open a wallet at the start of the day, and close it at the end of the day (using Alt + F4), I get a "Closing wallet" loading screen that never goes away. I have to alt-f4 again to close the UI but I still see the process running in Task Manager
![image](https://user-images.githubusercontent.com/21298211/153806551-6790c259-5be7-4688-a759-f2ed62a4ed35.png)

This issue cannot be reproduced by simply opening and closing the wallet. It happens if I have the wallet open for hours. It may be able to reproduce if you have the wallet open for 10 minutes, but I'm not sure.

Furthermore, when I log out the wallet, I get presented with the Welcome to Monero screen instead of the select wallet screen. I know it's a separate issue, but usually when people log out of one wallet, it's to open the other wallet, not to create a new wallet or something, because that's a one time thing, whereas the open another wallet happens everytime after.

Another feature request is to add a wallet_name[view-only].monero_gui file to the wallet directories that can be double clicked.

Another Bug, is clicking on monero: links opens a new instance of the GUI and doesn't even auto fill the data. What I expect is for the existing GUI to focus and auto fill the data or at least the new instance to autofill the data. I think you could use IPC to inform the existing instance of the data. In my application, I make a post request to 127.0.0.1 with the relevant data.

**Comment by elibroftw** [comment]
When decorations are off, the closing works (for now), but try to close the wallet also results in an infinite loop.
![image](https://user-images.githubusercontent.com/21298211/154772178-f7e4b6a1-609a-4bd1-8c23-7b16cc8e9091.png)

**Comment by selsta** [comment]
Just to confirm, is your wallet fully synced?

**Comment by selsta** [comment]
So pressing the X button works, but pressing the close button in Settings doesn't?

**Comment by selsta** [comment]
Thanks, that might give me a tip on where to look.

**Comment by elibroftw** [comment]
![image](https://user-images.githubusercontent.com/21298211/155064582-94c10055-7252-45ed-b0e0-3b558b1ed1c7.png)
It happened even with decorations off

**Comment by selsta** [comment]
Can you go to the main menu, then click on "Change wallet mode" and then enable the portable mode checkbox.

Afterwards can you check if you still have issues with closing getting stuck?

**Comment by elibroftw** [comment]
Issue still occurs even in portable mode

**Comment by mj-xmr** [comment]
@elibroftw : Does this also happen on an external node, or only local?

**Comment by mj-xmr** [comment]
So far I couldn't reproduce it in neither Win7 under a VM, nor native Win10. From my prior experience with GUIs, such symptoms result from threads in a deadlock. To continue on this issue, I will simply focus on examining how the threads are being destroyed upon exit and simulate failures there. In fact, during creation of my autogenerated reports I experienced such deadlocks as well.

@elibroftw : May I know your network speed in Mb/s? I'm trying to find clues.

**Comment by elibroftw** [comment]
50

**Comment by selsta** [comment]
@devhyper did you ever run into this issue on Windows?

**Comment by devhyper** [comment]
I don't think I experienced this, but I'll check again later today.

**Comment by selsta** [comment]
Thanks! Having the GUI open for a while helps with reproducing.

**Comment by MajesticBank** [comment]
I have same issue happening on Debian in Virtual machine. 

GUI didn't want to cooperate displaying Closing wallet after trying to close the GUI with X, after stopping monerod it changed to "Checking local node status"

Since I couldn't restart that Virtual machine I've moved it to another Desktop and it's still spinning around Monero logo for 3 days.

It does not impact working of monerod and using that wallet from monero-cli / monero-rpc.

**Comment by selsta** [comment]
@MajesticBank still stuck at "Checking local node status"? Might be a different thing then.

**Comment by MajesticBank** [comment]
Yes,  it just hang on that "Checking local node status" and like every 30-45 seconds I see spike of cpu usage on that GUI process.

Not sure how reproducible it might be

**Comment by devhyper** [comment]
Sorry for the delay, I was away this weekend and completely forgot.

I was able to reproduce the issue, the GUI was open for around 5 hours.

**Comment by selsta** [comment]
@devhyper Any idea how to debug it? I tried WinDbg but wasn't even able to get a simple backtrace like in gdb.

**Comment by elibroftw** [comment]
We should think about it holistically. I.e which qml file is responsible for showing the loader -> which variable controls the loader showing -> when is the variable supposed to change -> when does the variable actually change.

This can avoid using the debugger. Unless of course this issue is a bug in a library.

**Comment by mj-xmr** [comment]
> This can avoid using the debugger. Unless of course this issue is a bug in a library.

Yes. I was able to solve a hard problem like this without any debugger, by simply bisecting the active lines of code. The luxury I was having then was that it was reproducible on my machine.

**Comment by elibroftw** [comment]
https://github.com/monero-project/monero-gui/blob/master/src/libwalletqt/WalletManager.cpp#L215

Either the mutex is broken here, the address(0,0) is broken, or async is broken. My order of presumptions is mutex, address(0,0), and then async might be broken.

**Comment by mj-xmr** [comment]
Thanks for the clue. I'll try to simulate these situations via proper tests.

**Comment by selsta** [comment]
I need someone who can use this binary and reproduce the close bug: https://github.com/selsta/monero-gui/suites/5993379160/artifacts/206729904

It will open a cmd window in addition to the GUI and print some debug info so we can check where exactly it gets stuck. I can't reproduce the issue consistently in my VM.

@elibroftw maybe?

**Comment by elibroftw** [comment]
I don't see any output in the command line during the infinite loop.

**Comment by selsta** [comment]
I need the output before the infinite loop.

**Comment by elibroftw** [comment]
```
2022-04-09 15:46:39.407 W Retrying to obtain clipboard.
2022-04-09 15:52:32.324 W Retrying to obtain clipboard.
2022-04-09 16:26:23.062 W Retrying to obtain clipboard.
<infinite loop happens just before next line>
2022-04-09 16:26:23.815 W Retrying to obtain clipboard
```

**Comment by selsta** [comment]
Does it print `WalletManager::closeWallet() start` anywhere?

**Comment by elibroftw** [comment]
no

**Comment by selsta** [comment]
You might have to set log level in Settings -> Log to 2. Then test by closing the wallet (not the GUI) and check if you see messages like `WalletManager::closeWallet() start`.

If yes try to reproduce the infinite loop issue.

**Comment by selsta** [comment]
<img width="985" alt="Screenshot 2022-04-09 at 18 28 53" src="https://user-images.githubusercontent.com/7697454/162582808-494d6cea-38bf-4773-a799-e45306673991.png">

Closing the wallet should look like this.

**Comment by elibroftw** [comment]
```
2022-04-09 18:52:55.971 D Displaying processing splash
2022-04-09 18:52:55.974 I WalletManager::closeWalletAsync() start
2022-04-09 18:52:55.975 I WalletManager::closeWalletAsync() end
2022-04-09 18:52:56.481 D Some problems at connect, message: No connection could be made because the target machine actively refused it
2022-04-09 18:52:56.481 D Failed to connect to localhost:18081
2022-04-09 18:52:56.482 I Failed to invoke http request to  /mining_status
2022-04-09 18:52:56.482 D Reconnecting...
2022-04-09 18:52:58.456 D Already seen <...>
```

**Comment by elibroftw** [comment]
I guess this is an async issue then?

**Comment by selsta** [comment]
> I guess this is an async issue then?

Yes.

**Comment by elibroftw** [comment]
you got any ideas what to do from here? I don't have a clue what the scheduler is.

**Comment by selsta** [comment]
One thing to test is to increase the maxThreadCount here: https://github.com/monero-project/monero-gui/pull/3198/

That's not really a fix but it would be interesting to see if you can still reproduce the issue. Will make a binary for it.

**Comment by elibroftw** [comment]
Compare https://www.qt.io/blog/2017/04/18/multithreaded-programming-future-promise to https://github.com/monero-project/monero-gui/blob/master/src/libwalletqt/WalletManager.cpp#L229
and https://github.com/monero-project/monero-gui/blob/master/src/qt/FutureScheduler.cpp#L32. I don't know about you but the blog is way easier to understand. Under "Make the example more complicated" is a callback example. Is this something useful?

**Comment by selsta** [comment]
This is the binary with macThreadCount set to 6 instead of 4: https://github.com/selsta/monero-gui/suites/6059803194/artifacts/209160197

**Comment by elibroftw** [comment]
No change. What were your thoughts on that blog post I posted?

**Comment by selsta** [comment]
I did look through it but it wasn't helpful yet to me.

I have to add more logging the scheduler to see where exactly it fails.

Maybe then I can compare it with the example you linked, or search for other examples and see if something is different or wrong.

**Comment by selsta** [comment]
@elibroftw I'm sorry for asking you to test a new binary again: https://github.com/selsta/monero-gui/suites/6061261400/artifacts/209250833

I added a lot of logging to the scheduler.

**Comment by elibroftw** [comment]
no worries happy to help.

**Comment by elibroftw** [comment]
```
2022-04-11 03:16:45.994 I WalletManager::closeWalletAsync() start
2022-04-11 03:16:45.995 I FutureScheduler::run(callback) start
2022-04-11 03:16:45.995 I FutureScheduler::add() start
2022-04-11 03:16:45.996 I FutureScheduler::add() after locker mutex
2022-04-11 03:16:45.996 I FutureScheduler::add() increment alive
2022-04-11 03:16:45.996 I FutureScheduler::add() end
2022-04-11 03:16:45.997 I FutureScheduler::run(callback) inside first return
2022-04-11 03:16:45.997 I WalletManager::closeWalletAsync() end
2022-04-11 03:16:46.005 I FutureScheduler::stopping() start
2022-04-11 03:16:46.107 I FutureScheduler::stopping() start
2022-04-11 03:16:46.207 I FutureScheduler::stopping() start
2022-04-11 03:16:46.308 I FutureScheduler::stopping() start
2022-04-11 03:16:46.409 I FutureScheduler::stopping() start
2022-04-11 03:16:46.509 I FutureScheduler::stopping() start
2022-04-11 03:16:46.532 I FutureScheduler::run(no callback) start
2022-04-11 03:16:46.533 I FutureScheduler::add() start
2022-04-11 03:16:46.533 I FutureScheduler::add() after locker mutex
2022-04-11 03:16:46.534 I FutureScheduler::add() increment alive
2022-04-11 03:16:46.534 I FutureScheduler::add() end
2022-04-11 03:16:46.534 I FutureScheduler::run(no callback) inside first return
2022-04-11 03:16:46.610 I FutureScheduler::stopping() start
2022-04-11 03:16:46.712 I FutureScheduler::stopping() start
2022-04-11 03:16:46.817 I FutureScheduler::stopping() start
2022-04-11 03:16:46.919 I FutureScheduler::stopping() start
2022-04-11 03:16:47.032 I FutureScheduler::stopping() start
2022-04-11 03:16:47.133 I FutureScheduler::stopping() start
2022-04-11 03:16:47.234 I FutureScheduler::stopping() start
2022-04-11 03:16:47.335 I FutureScheduler::stopping() start
```
I don't think this is helpful, but here you go.

**Comment by selsta** [comment]
Thank you, yeah unfortunately nothing interesting yet. Will try to make the close wallet function non async as a workaround, let's see if that works.

**Comment by selsta** [comment]
https://github.com/monero-project/monero-gui/pull/3885 this seems to work for me locally even better than before, seems like there was no reason to async close in the first place.

Closing is way faster now.

**Comment by selsta** [comment]
@elibroftw Please test https://github.com/monero-project/monero-gui/suites/6082784802/artifacts/210668099 it should have the close bug fixed.

I mean I still don't know the underlying issue with FutureScheduler but until I can reproduce it locally I don't really know how to debug it.

That reminds me that you had issues with the fee estimate, maybe that's also related to this scheduler bug...

**Comment by elibroftw** [comment]
![image](https://user-images.githubusercontent.com/21298211/163035253-9685c91f-710b-4639-a0c1-850a5b7608df.png)
🔥

**Comment by selsta** [comment]
Did it get stuck? I can't tell from the screenshot :S

**Comment by elibroftw** [comment]
Yep it got stuck

**Comment by selsta** [comment]
Does this happen every time you close it or do you have to try a couple times?

**Comment by selsta** [comment]
Everything works perfectly in my VM Windows :/

**Comment by elibroftw** [comment]
Only after waiting a couple hours. If I open and close immediately, it closes really fast. So I'll wait an hour now.

**Comment by selsta** [comment]
ok, trying that now

**Comment by selsta** [comment]
Had it open for almost 20 hours, no issues hmm

**Comment by selsta** [comment]
sure

**Comment by selsta** [comment]
https://github.com/selsta/monero-gui/suites/6111477246/artifacts/212385368

**Comment by selsta** [comment]
ping @elibroftw any luck?

**Comment by elibroftw** [comment]
I'll try now.

**Comment by elibroftw** [comment]
```
2022-04-21 23:15:21.120 D Reconnecting...
2022-04-21 23:15:21.286 D SSL handshake success
2022-04-21 23:15:21.427 I FutureScheduler::run(callback) before done
2022-04-21 23:15:21.427 I FutureScheduler::done() start
2022-04-21 23:15:21.428 I FutureScheduler::done() before locker mutex
2022-04-21 23:15:21.428 I FutureScheduler::done() after locker mutex
2022-04-21 23:15:21.428 I FutureScheduler::done() before wakeAll
2022-04-21 23:15:21.429 I FutureScheduler::done() end
2022-04-21 23:15:21.429 I FutureScheduler::run(callback) after done
2022-04-21 23:15:21.429 I FutureScheduler::run(no callback) inside second return
2022-04-21 23:15:21.430 I FutureScheduler::run(no callback) calling function
2022-04-21 23:15:21.430 I FutureScheduler::run(no callback) start
2022-04-21 23:15:21.430 I FutureScheduler::add() start
2022-04-21 23:15:21.430 I FutureScheduler::add() after locker mutex
2022-04-21 23:15:21.431 I FutureScheduler::add() inside if (Stopping)
2022-04-21 23:15:21.431 I FutureScheduler::run(no callback) before done
2022-04-21 23:15:21.431 I FutureScheduler::done() start
2022-04-21 23:15:21.432 I FutureScheduler::done() before locker mutex
2022-04-21 23:15:21.432 I FutureScheduler::done() after locker mutex
2022-04-21 23:15:21.432 I FutureScheduler::done() before wakeAll
2022-04-21 23:15:21.433 I FutureScheduler::done() end
2022-04-21 23:15:21.433 I FutureScheduler::run(no callback) after done
2022-04-21 23:15:21.433 I FutureScheduler::shutdownWaitForFinished inside while
2022-04-21 23:15:21.434 I FutureScheduler::run(no callback) inside second return
2022-04-21 23:15:21.434 I FutureScheduler::run(no callback) calling function
2022-04-21 23:15:22.642 D Some problems at connect, message: No connection could be made because the target machine actively refused it
2022-04-21 23:15:22.643 D Failed to connect to localhost:18081
2022-04-21 23:15:22.644 I Failed to invoke http request to  /mining_status
2022-04-21 23:15:22.644 I FutureScheduler::run(no callback) before done
2022-04-21 23:15:22.644 I FutureScheduler::done() start
2022-04-21 23:15:22.645 I FutureScheduler::done() before locker mutex
2022-04-21 23:15:22.645 I FutureScheduler::done() after locker mutex
2022-04-21 23:15:22.645 I FutureScheduler::done() before wakeAll
2022-04-21 23:15:22.646 I FutureScheduler::done() end
2022-04-21 23:15:22.646 I FutureScheduler::run(no callback) after done
2022-04-21 23:15:22.646 I FutureScheduler::run(no callback) inside second return
2022-04-21 23:15:22.646 I FutureScheduler::run(no callback) calling function
```

**Comment by selsta** [comment]
Then it gets stuck? Also can you post the logs from a bit earlier?

**Comment by elibroftw** [comment]
Updated just now

**Comment by selsta** [comment]
The last thing I see is the wallet destructor.

**Comment by elibroftw** [comment]
Totally forgot about ~, although how do we know that the destructor finished running? The log just says closing not closed.

**Comment by selsta** [comment]
Yes, it gets stuck inside the destructor. I have to add more logging to it, and I will remove the async logging.

**Comment by elibroftw** [comment]
Watch this be a msys bug

**Comment by selsta** [comment]
It prints `qInfo() << "WalletManager::closeWallet() after m_currentWallet delete";` after the destructor

**Comment by elibroftw** [comment]
This is an issue in monero-core, not monero-gui no?

**Comment by elibroftw** [comment]
nvm, https://github.com/monero-project/monero-gui/blob/master/src/libwalletqt/Wallet.cpp#L1157

**Comment by elibroftw** [comment]
There is something wrong with this mutex
https://github.com/monero-project/monero-gui/blob/master/src/qt/FutureScheduler.cpp#L23 
https://github.com/monero-project/monero-gui/blob/master/src/qt/FutureScheduler.h#L62

**Comment by elibroftw** [comment]
Is there any way to figure what what has the mutex locked?

**Comment by elibroftw** [comment]
Or wait, whatever is supposed to make Alive > 0 is not working.

**Comment by elibroftw** [comment]
Hmm interesting, https://github.com/monero-project/monero-gui/blob/master/src/qt/FutureScheduler.cpp#L94. Therefore, there's another background task that doesn't want to stop.

**Comment by selsta** [comment]
I'll try to add logging in a way that we can see which function is involved.

**Comment by selsta** [comment]
@elibroftw One thing you can quickly test is if you can still reproduce the bug if you have the fiat price ticker disabled.

**Comment by elibroftw** [comment]
No difference.

**Comment by selsta** [comment]
Ok, then back to more logging.

**Comment by selsta** [comment]
@elibroftw more detailed logging: https://github.com/selsta/monero-gui/suites/6214720654/artifacts/219787383

**Comment by elibroftw** [comment]
```
2022-04-22 19:35:32.262 D Displaying processing splash
2022-04-22 19:35:32.265 I WalletManager::closeWallet() start
2022-04-22 19:35:32.266 I WalletManager::closeWallet() after locker
2022-04-22 19:35:32.266 I WalletManager::closeWallet() if (m_currentWallet == true)
2022-04-22 19:35:32.267 I WalletManager::closeWallet() after address result
2022-04-22 19:35:32.267 D ~Wallet: Closing wallet
2022-04-22 19:35:32.268 D ~Wallet: walletImpl -> stop
2022-04-22 19:35:32.268 I "~Wallet"   FutureScheduler::shutdownWaitForFinished start
2022-04-22 19:35:32.269 I "~Wallet"   FutureScheduler::shutdownWaitForFinished inside while
2022-04-22 19:35:47.254 D Already seen <5807601b7ceecb60402e32ed4ff9d693a3c279de3534c788b0c47f5b4dff260a>, and not for us, skipped
2022-04-22 19:35:47.255 D Already seen <7f70bb50977e29fbd3abbd624875f3c367c27eee82a6567638ef039cc6f1c911>, and not for us, skipped
2022-04-22 19:35:47.256 D Already seen <41bc809803c51b388fa172cc6c170dd006cad84f539f9cd052cfbafa4f2c8f33>, and not for us, skipped
2022-04-22 19:35:47.256 D Already seen <be11dee162b3fdc90117934d826a174296d7c1435bc7926cb5e398462150d247>, and not for us, skipped
2022-04-22 19:35:47.256 D Already seen <ced7010112d881ac15ac53bd924337c3339d5dc5aa5e017f3162825af1d46a51>, and not for us, skipped
2022-04-22 19:35:47.257 D Already seen <435ea99648933148973b3dee9165c6417306924e833082cd386d2c6e5d901d68>, and not for us, skipped
2022-04-22 19:35:47.257 D Already seen <86806b246cec3d2eea367bde363956e3def49056e424799c55488abfe43d0876>, and not for us, skipped
2022-04-22 19:35:47.257 D Already seen <1d2df45419ff4bdbf596b70c3f70e34ea41099cb169c355effc8c6d63aff0a95>, and not for us, skipped
2022-04-22 19:35:47.258 D Already seen <fc9c0ebeca708c7448cabedbd40cf14ce0f5c38e4ac26e7665da4b6e9b9632a5>, and not for us, skipped
2022-04-22 19:35:47.258 I Found new pool tx: <df70849a3e7216dd7a32747b0f486befbabe4efaa192a74a56d8d8da11446da5>
2022-04-22 19:35:47.258 I Found new pool tx: <395f3932cbd30ca96918c7e99ec19e03c4fcbfa69e24daf54a80b0c955fa50a7>
2022-04-22 19:35:47.259 I Found new pool tx: <6b3a2a3e64189afca50ec7a3827e4de36eadab8c83487eee6e692b4b67a655b2>
2022-04-22 19:35:47.259 D Already seen <4c5f3b5241a39aa08ba4814e1d2551bfcf3a73a9659038010c6ecf9fb53522c4>, and not for us, skipped
2022-04-22 19:35:47.259 D Already seen <7e288ce570d674824cf10cf8bf2049a2dbf704cbdada506d1bd32548636f5edd>, and not for us, skipped
2022-04-22 19:35:47.260 D Already seen <849a112a08ff26229888a388ed8752e05cbb59c4b89431af55accb3572ad91dd>, and not for us, skipped
2022-04-22 19:35:47.260 D Already seen <00daa3d56e68993ceab73655e70398db3b7cd3ad10775220f996dfadc41dedf5>, and not for us, skipped
2022-04-22 19:35:47.261 D asking for 3 transactions
2022-04-22 19:35:47.380 D Got 1 and OK
2022-04-22 19:35:47.381 I Refresh done, blocks received: 0, balance (all accounts): 11.852265634009, unlocked: 11.852265634009
2022-04-22 19:35:47.381 D refreshed
2022-04-22 19:35:47.482 I "startRefreshThread"   FutureScheduler::stopping() start
2022-04-22 19:35:47.483 I "Wallet startRefreshThread"   FutureScheduler::run(no callback) before done
2022-04-22 19:35:47.483 I "Wallet startRefreshThread"   FutureScheduler::done() start
2022-04-22 19:35:47.484 I "Wallet startRefreshThread"   FutureScheduler::done() before locker mutex
2022-04-22 19:35:47.484 I "Wallet startRefreshThread"   FutureScheduler::done() after locker mutex
2022-04-22 19:35:47.484 I "Wallet startRefreshThread"   FutureScheduler::done() before wakeAll
2022-04-22 19:35:47.485 I "Wallet startRefreshThread"   FutureScheduler::done() end
2022-04-22 19:35:47.485 I "~Wallet"   FutureScheduler::shutdownWaitForFinished end
2022-04-22 19:35:47.485 D trimming to 2230561, offset 2230561
2022-04-22 19:35:47.486 I "Wallet startRefreshThread"   FutureScheduler::run(no callback) after done
2022-04-22 19:35:47.486 D trimming to 2230561, offset 2230561
2022-04-22 19:35:47.552 D Wallet cache stored successfully
2022-04-22 19:35:47.553 I ~WalletImpl
2022-04-22 19:35:47.553 D pauseRefresh: refresh paused...
2022-04-22 19:35:47.554 I closing wallet...
2022-04-22 19:35:47.555 I Calling wallet::stop...
2022-04-22 19:35:47.555 I wallet::stop done
2022-04-22 19:35:47.556 I ~WalletImpl finished
2022-04-22 19:35:47.557 D Problems at ssl shutdown: uninitialized
2022-04-22 19:35:47.558 D Problems at cancel: The file handle supplied is not valid
2022-04-22 19:35:47.558 D Problems at shutdown: The file handle supplied is not valid
2022-04-22 19:35:47.592 D m_walletImpl deleted
2022-04-22 19:35:47.593 I FutureScheduler::~FutureScheduler start
2022-04-22 19:35:47.594 I "~FutureScheduler"   FutureScheduler::shutdownWaitForFinished start
2022-04-22 19:35:47.594 I "~FutureScheduler"   FutureScheduler::shutdownWaitForFinished end
2022-04-22 19:35:47.594 I FutureScheduler::~FutureScheduler end
2022-04-22 19:35:47.595 I WalletManager::closeWallet() after m_currentWallet delete
2022-04-22 19:35:47.595 I WalletManager::closeWallet() end
2022-04-22 19:35:47.596 D Hiding processing splash
```
This happened on my first try. It took 15 seconds to close. 
Subsequent tries were smooth so I'll leave the wallet running overnight to see if the trigger is longer now.

**Comment by selsta** [comment]
Did you close it immediately after starting? From the logs it was still doing wallet refresh in the background.

15 seconds is long but at least it worked in the end.

**Comment by elibroftw** [comment]
No I closed it after hours. I left it running overnight and it still managed to close. I think it's a good fix then for now.

**Comment by selsta** [comment]
I'm not really sure why you can't reproduce the issue now.

I only have extra logging plus #3889 and #3890 but they aren't expected to fix this bug.

**Comment by elibroftw** [comment]
Maybe both of them together fix the issue?

**Comment by selsta** [comment]
3889 is only relevant when using a local node and 3890 is only relevant when having fiat price enabled. I think you tested without these conditions.

**Comment by elibroftw** [comment]
I saw the mining status messages though in the log before even with a remote node https://github.com/monero-project/monero-gui/issues/3840#issuecomment-1094106293. Or am I mistaken?

**Comment by selsta** [comment]
No idea where that comes from again :/

**Comment by selsta** [comment]
So what do you think? Should we keep async during close or remove it? I think async is the more correct way but it appears to be slower.

Should I make a binary with master + 3889 + 3890 so that we can double confirm that the bug is fixed?

**Comment by elibroftw** [comment]
Yeah make that and make one with async as well so I can see which has the best experience.

**Comment by selsta** [comment]
Async: https://github.com/selsta/monero-gui/suites/6233216680/artifacts/221206532

Will make another one for without async.

**Comment by selsta** [comment]
Without async: https://github.com/selsta/monero-gui/suites/6234076301/artifacts/221248748

**Comment by elibroftw** [comment]
Without async is faster and not to mention that if something goes wrong, we'll see a UI freeze rather than an infinite loop.

**Comment by selsta** [comment]
Just to confirm, both versions have the close bug fixed?

**Comment by elibroftw** [comment]
Yes

**Comment by selsta** [comment]
that's frustrating because I don't know what fixed it :D

**Comment by elibroftw** [comment]
I'm running it on my laptop now to double check if it wasn't my PC settings that fixed something.

**Comment by elibroftw** [comment]
Works fine on laptop too. I suspect that mining status is to blame for most of this plus the fiat price fetching since I had that enabled too.

**Comment by selsta** [comment]
@MoneroArbo Can you also confirm if https://github.com/monero-project/monero-gui/issues/3840#issuecomment-1107659951 solved the close bug for you?

**Comment by MoneroArbo** [comment]
> @MoneroArbo Can you also confirm if [#3840 (comment)](https://github.com/monero-project/monero-gui/issues/3840#issuecomment-1107659951) solved the close bug for you?

Wow the new user flow for this version is so much better than the last official release, amazing!

And yeah, it seems not to hang on close either. Made a new wallet, opened an old wallet, opened and closed wallets a few times, with and without Window decorations, even closed it mid scan, all seems to work.

Coincidentally (or not) it also seems to no longer do the thing where the [wallet scan gets stuck at 0 blocks remaining](https://github.com/monero-project/monero-gui/issues/3594#issuecomment-870015644).

**Comment by selsta** [comment]
ok, thanks for testing! guess I can finally close this
